### PR TITLE
NuGetUtils update

### PR DESCRIPTION
**Pull request by asbjornu**
Rebase of #1634. Resolves #1634.

### CODE REVIEW

**Comment by asbjornu** [comment]
This is puzzling. This PR is just a rebase of #1634 (so all the merge commits are removed), but the build fails on this while it's successful in #1634. Any ideas why, @arturcic?

**Comment by arturcic** [comment]
Looks like everywhere the same test is failing OutputsShouldMatchVariableProvider

**Comment by asbjornu** [comment]
Any ideas why that fails here and not in #1634, @arturcic? I don't get it.

**Comment by arturcic** [comment]
Not really, we need to check if there is any difference between these 3 branches

**Comment by asbjornu** [comment]
@arturcic, I did a fresh `git reset` and `git rebase` now and that seems to have fixed the problem. Things are at least green. 🤷‍♂️

**Comment by arturcic** [comment]
Great! My question is why do we have 2 PRs?

**Comment by asbjornu** [comment]
Just to get rid of these: c2f36105a1ccf7f4ec13f8e97ec3743bfae85bab, 9d8eb3806d868300a92382bf2ea82ea4eb1b2641, b608f96175e90c5dbc8cddaa371c6439ad2ff3f7, 854f16b6753d84214c78af52224f90500da7ac11.

**Comment by arturcic** [comment]
So I guess this PR will get merged instead of the other one

**Comment by asbjornu** [comment]
Yes, this resolves #1634 so as soon this is merged, it will be closed automatically. It would be neat if GitHub supported `rebase` instead of `merge` for the "Update branch" button, but alas it doesn't. I've just installed [Autorebase](https://github.com/tibdex/autorebase) which might help us with that. I'll try that on #1634 shortly.

**Comment by asbjornu** [comment]
Sigh. Seems like my rebase failed for some reason.

File: src/GitVersionTask/GitVersionTask.csproj [hunk 1]
```diff
@@ -37,9 +38,7 @@
     <PackageReference Include="YamlDotNet" Version="$(PackageVersion_YamlDotNet)">
         <PrivateAssets>All</PrivateAssets>
     </PackageReference>
-    <!--See https://github.com/Microsoft/msbuild/issues/3671 for reason for the ExcludeAssets against msbuild packagerefernces -->
-    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.9.20" ExcludeAssets="runtime" />
-    <PackageReference Include="UtilPack.NuGet.MSBuild" Version="$(PackageVersion_UtilPackNuGetMSBuild)">
+    <PackageReference Include="NuGetUtils.MSBuild.Exec" Version="$(PackageVersion_UtilPackNuGetMSBuild)">
```
Comment (dazinator) [inline src/GitVersionTask/GitVersionTask.csproj]: @asbjornu @arturcic  This looks like a problem with the rebase. If you compare this csproj file to @stazz's PR you can see his is different: https://github.com/GitTools/GitVersion/pull/1634/files#diff-bca0d1ca1ea8fbf39936384490e12d86R41

**Comment by asbjornu** [comment]
Thanks for figuring that out, @dazinator. I think I got the rebase right now, but notice that 3b28212a117c8c51cd4f24b28f354afccc0ed2fc (and my rebased de3112488b7ea63ad4f34b13e3d0da38e92cb8c1) reintroduces the `Logger.Reset()` calls that @tpaxatb removed in #1644). I suppose that's a mistake, @stazz?

**Comment by tpaxatb** [comment]
> Thanks for figuring that out, @dazinator. I think I got the rebase right now, but notice that [3b28212](https://github.com/GitTools/GitVersion/commit/3b28212a117c8c51cd4f24b28f354afccc0ed2fc) (and my rebased [de31124](https://github.com/GitTools/GitVersion/commit/de3112488b7ea63ad4f34b13e3d0da38e92cb8c1)) reintroduces the `Logger.Reset()` calls that @tpaxatb removed in #1644. I suppose that's a mistake, @stazz?

I'm going to assume that it's actually ok as is, since the logger.reset is actually paired with a setloggers call at the beginning of the function (rather than in a constructor as it was previously...)

**Comment by asbjornu** [comment]
Thanks for the insight, @tpaxatb. Then I guess this just needs review and approval from @dazinator and we should be good to go. Do you agree, @arturcic?

**Comment by arturcic** [comment]
Agree

**Comment by stazz** [comment]
@asbjornu Yes, that is indeed a mistake. Rebase ended up being a bit messy even though I followed instructions, sorry about that. I have close to zero experience with rebasing stuff, as I've never worked with GIT repo which follows SVN workflow.

But yeah if @tpaxatb was ok with that, then it's fine. At least nothing should really interact directly with the MSBuild things anymore. The question is, does the NuGetUtils.MSBuild.Exec task factory handle logging using `IBuildEngine` in some special way? I think no, but maybe @tpaxatb can shed some insight on this? Currently the task factory is using the `IBuildEngine` provided in `Initialize` method when it performs pre-checks, and the task proxy will just redirect all output of invoked task to the `IBuildEngine` that it receives.

@dazinator Do I need to do any changes anywhere? The change request about the .csproj file seems to be not-relevant anymore, at least when looking at current state of https://github.com/asbjornu/GitVersion/blob/feature/NuGetUtilsUpdate/src/GitVersionTask/GitVersionTask.csproj .

**Comment by stazz** [comment]
Hmm, actually, it seems I have some bug introduced in recent commits, despite the fact that I've added tests for the NuGetUtils.MSBuild.Exec. I'll investigate and back to you.

**Comment by stazz** [comment]
> 
> 
> Hmm, actually, it seems I have some bug introduced in recent commits, despite the fact that I've added tests for the NuGetUtils.MSBuild.Exec. I'll investigate and back to you.

Found the bug, it was only appearing in some quite special circumstances (e.g. having multiple assembly files in same package or having certain dependencies). I've fixed it now for `2.0.5`, and updated [my fork](https://github.com/stazz/GitVersion/tree/stazz/NuGetUtilsUpdate).

**Comment by feitzi** [comment]
any updates?

**Comment by stazz** [comment]
Hey @feitzi , for my part, everything should be ok, I think. The NuGetUtils task factory has been tested and put into use by GitVersionTask. I am not sure what the conflicts mentioned here are exactly about - it seems that GitVersionTask is using workflow from SVN originally (based on some message I saw in this or another issue).
@dazinator @asbjornu if I need to still do something, feel free to instruct what needs to be done. :)

File: src/GitVersionTask/GenerateGitVersionInformation.cs [hunk 2]
```diff
@@ -1,42 +1,99 @@
 namespace GitVersionTask
 {
+    using System;
     using System.IO;
     using GitVersion;
     using GitVersion.Helpers;
-    using Microsoft.Build.Framework;
 
-    public class GenerateGitVersionInformation : GitVersionTaskBase
+    public static class GenerateGitVersionInformation
```
Comment (asbjornu) [inline src/GitVersionTask/GenerateGitVersionInformation.cs]: @stazz, why the change to `static class` here?

File: src/GitVersionTask/GenerateGitVersionInformation.cs [hunk 3]
```diff
@@ -1,42 +1,99 @@
...
+            };
+            var generator = new GitVersionInformationGenerator( fileWriteInfo.FileName, fileWriteInfo.WorkingDirectory, versionVariables, new FileSystem());
+            generator.Generate();
 
-            var workingDirectory = IntermediateOutputPath ?? TempFileTracker.TempPath;
+            return output;
+        }
 
-            GitVersionInformationFilePath = Path.Combine(workingDirectory, fileName);
+        public sealed class Input
```
Comment (asbjornu) [inline src/GitVersionTask/GenerateGitVersionInformation.cs]: @stazz, why isn't any form of inheritance used across the different `Input` classes (and `Output` classes, for that matter)?

File: src/GitVersionTask/GenerateGitVersionInformation.cs [hunk 4]
```diff
@@ -1,42 +1,99 @@
...
+            public string ProjectFile { get; set; }
+
+            public string IntermediateOutputPath { get; set; }
+
+            public string Language { get; set; }
+
+            public bool NoFetch { get; set; }
+        }
+
+        private static Boolean ValidateInput( this Input input )
```
Comment (asbjornu) [inline src/GitVersionTask/GenerateGitVersionInformation.cs]: @stazz, why isn't this method placed inside the `Input` class itself?

File: src/GitVersionTask/GetVersion.cs [hunk 5]
```diff
@@ -1,108 +1,129 @@
 namespace GitVersionTask
 {
-    using Microsoft.Build.Framework;
+    using System;
+    using GitVersion;
 
-    public class GetVersion : GitVersionTaskBase
+    public static class GetVersion
```
Comment (asbjornu) [inline src/GitVersionTask/GetVersion.cs]: @stazz, why the change to `static` here?

**Comment by stazz** [comment]
The `static` is because there is no longer need to extend `ITask`. Instead, the entrypoint acts just like a normal C# entrypoint, with an addition of having complex typed parameters (and not just `String[]`). That pretty much eliminates the need to use class hierarchy for the tasks contained in the assembly as well.

So did I undestand you correctly @asbjornu that are you suggesting that I make a fresh fork out of the latest tip of GitVersion repository, and re-apply changes for NuGetUtils? That makes sense for me, as I have a feeling that the change log is now way too convoluted. The work load of doing so is not too big, IMO, as the changes were relatively straightforward, but I think I'll have time for that only on this Sunday.

With the full re-apply of changes, I also hope it will be easier to see the reasoning behind the resulting `static` entrypoints and no class hierarchy for tasks.

File: src/GitVersionTask/GenerateGitVersionInformation.cs (hunk 2 above)
Comment (stazz) [reply src/GitVersionTask/GenerateGitVersionInformation.cs]: The reason is that the class has no state on its own anymore (instead, the `Input` and `Output` inner classes are used), so no instances of it will be generated either.

File: src/GitVersionTask/GenerateGitVersionInformation.cs (hunk 3 above)
Comment (stazz) [reply src/GitVersionTask/GenerateGitVersionInformation.cs]: I simply forgot to do that. I'll fix this on Sunday.

File: src/GitVersionTask/GetVersion.cs (hunk 5 above)
Comment (stazz) [reply src/GitVersionTask/GetVersion.cs]: Just like `GenerateGitVersionInformation` (and one other task too), this class has no state on its own anymore, so no instances of it will be created either.

File: src/GitVersionTask/GenerateGitVersionInformation.cs (hunk 4 above)
Comment (stazz) [reply src/GitVersionTask/GenerateGitVersionInformation.cs]: No particular reason. I guess the validity input depends a bit on a context (which I admit in this case is exactly one). Can move it as instance method of `Input` class on Sunday.

File: src/GitVersionTask/GenerateGitVersionInformation.cs (hunk 2 above)
Comment (asbjornu) [reply src/GitVersionTask/GenerateGitVersionInformation.cs]: I see. Thanks for the explanation.

File: src/GitVersionTask/GenerateGitVersionInformation.cs (hunk 3 above)
Comment (asbjornu) [reply src/GitVersionTask/GenerateGitVersionInformation.cs]: Great!

**Comment by asbjornu** [comment]
@stazz:

>So did I undestand you correctly @asbjornu that are you suggesting that I make a fresh fork out of the latest tip of GitVersion repository, and re-apply changes for NuGetUtils?

Yes, exactly.

>That makes sense for me, as I have a feeling that the change log is now way too convoluted. The work load of doing so is not too big, IMO, as the changes were relatively straightforward, but I think I'll have time for that only on this Sunday.

I'm very happy to hear that! Awesome!

>With the full re-apply of changes, I also hope it will be easier to see the reasoning behind the resulting `static` entrypoints and no class hierarchy for tasks.

👍

**Comment by stazz** [comment]
@asbjornu This is now opened as #1677 . :)

**Comment by asbjornu** [comment]
Resolved by #1677.
